src/Mapper.php

```diff
@@ -1,11 +1,156 @@
+use Tnapf\JsonMapper\Attributes\StringType;
 
 class Mapper implements MapperInterface
 {
+    protected ?CaseConversionInterface $caseConversion = null;
+    protected object $instance;
+    protected ReflectionClass $reflection;
+
+    /**
+     * @var BaseType[]
+     */
+    protected array $attributes;
+
+    protected string $class;
+    protected array $data;
+
     public function map(string $class, array $data): object
     {
-        $instance = new $class();
+        $instance = (new self());
+        $instance->class = $class;
+        $instance->data = $data;
+        $instance->reflection = new ReflectionClass($class);
+
+        if ($attributes = $instance->reflection->getAttributes(CaseConversionInterface::class, ReflectionAttribute::IS_INSTANCEOF)) {
+            if (count($attributes) > 1) {
+                throw new MapperException("{$class} has more than one case conversion attribute");
+            }
+
+            $instance->caseConversion = $attributes[0]->newInstance();
+        }
+
+        $instance->instance = new $class();
+
+        return $instance->doMapping();
+    }
+
+    protected function convertNameToCase(string $name): string
+    {
+        return $this?->caseConversion?->convertToCase($name) ?? $name;
+    }
+
+    protected function convertNameFromCase(string $name): string
+    {
+        return $this?->caseConversion?->convertFromCase($name) ?? $name;
+    }
+
+    protected function doMapping(): object
+    {
+        $this->fillPropertyAttributes();
+
+        foreach ($this->attributes as $types) {
+            $attribute = $types[0];
+            $data = $this->data[$this->convertNameFromCase($attribute->name)] ?? null;
+
+            if ($data === null) {
+                if ($attribute->nullable) {
+                    continue;
+                }
+
+                throw new MapperException("Property {$attribute->name} is not nullable");
+            }
+
+            $validType = false;
+
+            foreach ($types as $attribute) {
+                if ($attribute instanceof ObjectType) {
+                    $data = $this->map($attribute->class, $data);
+                }
+
+                if ($attribute instanceof ObjectArrayType) {
+                    $data = array_map(
+                        fn ($item) => $this->map($attribute->class, $item),
+                        $data
+                    );
+                }
+
+                if ($attribute->isType($data)) {
+                    $validType = true;
+                    break;
+                }
+            }
+
+            if (!$validType) {
+                throw new MapperException(
+                    "Property {$attribute->name} is not of type ".
+                    implode(
+                        ', ',
+                        array_map(static fn ($type) => $type::class, $types)
+                    )
+                );
+            }
+
+            $this->instance->{$this->convertNameToCase($attribute->name)} = $data;
```

Comment:
Property may not be accessible. Use `ReflectionProperty::setValue()` to avoid errors on non-accessible properties.

Reply:
I grab all public properties to prevent this

Reply:
Yeah, but that's weird. What if I want to make a readonly mapped instance? I would make private properties with getters to avoid modifications.

I guess this behaviour should be configurable and default to "all properties". Presence or lack of property attribute should hint at property being mappable.

Reply:
Issue is I don't have access to the ReflectionProperty in that scope.